### PR TITLE
#2478 MMU Jam and extruder grinding detection.

### LINKED ISSUE #2478: [ENHANCEMENT] MMU jam/grinding detection.
**Printer type** - MK3S
**MMU Upgrade** - MMU2S

For the -S variant with the IR sensor, the sensor state is ignored when printing. If there's a jam, the filament grinds and the IR sensor will change to zero. Seems useful to detect this and likely save the print (or at least the user's time). This should work for both heat-creep jams or feed-tube/buffer jams and binding. 

I haven't quite finished poking around in the mmu handling routines yet but I suspect the end result will probably be a jam notice that the mmu needs attention. Depending on the nature of the jam it may just make the situation more difficult if automatic action is attempted - I think it's best to let the user determine what's up. 

-> This will only work for the 3S variant; I hear the Mk3 sensor's reliability has improved significantly, but I suspect it's rarely used for the MMU anymore owing to the improved loading from using the IR lever.

It's been requested before in #2214 which was closed and noted it was in the backlog. I'm noting this as a new open issue because I'm going to take a stab at this; some of my filament seems more prone to this problem than others. Please feel free to add comments/concerns.

### PR DESCRIPTION
**Pull request by vintagepc**
Filing a PR since I think this is ready for more general review and consideration. 
This closes #2478 and the original #2214 
The theory here is to detect a jam/ground filament state by also monitoring the extruder filament sensor in addition to the FINDA. If grinding is detected, the MMU goes into "attention required" state to hopefully let the user fix the issue and resume the print, rather than continuing blindly and printing air for the remainder. 

This feature can be easily disabled via the Filament sensor setting item if false positives are a problem for certain filaments (e.g. flex). 

There are more implementation details and some test outlines/screen pics in #2478. 

Items TBD:
- Translations for new jam detection screen.
- ~Find a solution for the new F. Sensor menu choice being 1 char too long (some options proposed in linked issue, may require language/translator input)~ I could not come up with anything more succinct than "+Jam" to indicate it's on + jam detection mode.
- Verify normal behaviour without an MMU
- Verify behaviour on a MK3 (I only have a 3S)

### CODE REVIEW

**Comment by vintagepc** [comment]
Test case: I just encountered this "for real" - there was some hard junk in my filament that blocked the nozzle and caused the extruder to grind. Jam detection tripped and it recovered beautifully - there was only a minor blemish in the infill it was doing when it tripped. 

I know it's too late for 3.9 but I really hope you consider including this in 3.9.1 or 3.10

**Comment by vintagepc** [comment]
w00t!

**Comment by leptun** [comment]
@vintagepc You will probably end up splitting the option into two. So you have `F. Sensor On/Off` and `Jam detection On/Off` and maybe also `N/A` if the Fsensor is disabled. This way the option is shown to the user only when the MMU is connected and the text fits. I'll try to test the PR today on my MK3S with and without MMU

**Comment by vintagepc** [comment]
understood; Just say the word and I am happy to adjust.
(FWIW, I run a Mk3-based local branch that I merge my feature/PR branches into before compiling) so certainly on my own machine these sorts of things are under continuous test, not just once or twice during dev.)

**Comment by 3d-gussner** [comment]
@vintagepc Please review your PR as there are conflicts with latest firmware release.

File: Firmware/messages.c [hunk 1]
```diff
@@ -104,6 +104,7 @@ const char MSG_V2_CALIBRATION[] PROGMEM_I1 = ISTR("First layer cal."); ////c=17
 const char WELCOME_MSG[] PROGMEM_I1 = ISTR(CUSTOM_MENDEL_NAME " OK."); ////c=20
 const char MSG_OFF[] PROGMEM_I1 = ISTR("Off"); ////
 const char MSG_ON[] PROGMEM_I1 = ISTR("On"); ////
+const char MSG_ON_JAM[] PROGMEM_I1 = ISTR("+Jam"); ////
```
Comment (3d-gussner) [inline Firmware/messages.c]: Please add `c=xx` behind `; ////` where `xx` is the max. character length for this toggle.
Remember that the LCD line is limited to 20 chars, so the MSG and the TOGGLE MSG including brackets and other chars (>, spaces, ...) don't exceed the 20 chars.

File: Firmware/mmu.cpp [hunk 2]
```diff
@@ -829,7 +839,15 @@ void mmu_M600_wait_and_beep() {
 		KEEPALIVE_STATE(PAUSED_FOR_USER);
 
 		int counterBeep = 0;
-		lcd_display_message_fullscreen_P(_i("Remove old filament and press the knob to start loading new filament."));
+		if (mmu_E_jam_detected)
+		{
+			lcd_display_message_fullscreen_P(_i("Possible extruder jam detected (IR=0). Fix problem and press knob."));
```
Comment (3d-gussner) [inline Firmware/mmu.cpp]: Please add ` ////c=20 r=yy` behind `"));" to document the need rows on the LCD screen.
Also don't forget that translations are often longer than English.

File: Firmware/mmu.cpp [hunk 3]
```diff
@@ -829,7 +839,15 @@ void mmu_M600_wait_and_beep() {
...
+			mmu_E_jam_detected = false; // Clear flag, we'll proceed with normal M600 change from hereon.
+		}
+		else
+		{
+			lcd_display_message_fullscreen_P(_i("Remove old filament and press the knob to start loading new filament."));
```
Comment (3d-gussner) [inline Firmware/mmu.cpp]: See comment above

File: Firmware/ultralcd.cpp [hunk 4]
```diff
@@ -5199,7 +5215,14 @@ do\
...
     {\
         /* Filament sensor turned on, working, no problems*/\
-        MENU_ITEM_TOGGLE_P(_T(MSG_FSENSOR), _T(MSG_ON), lcd_fsensor_state_set);\
+		if (oFSensorMode==ClFSensorMode::_On_And_Jam && mmu_enabled)\
+		{\
+        	MENU_ITEM_TOGGLE_P(_T(MSG_FSENSOR), _T(MSG_ON_JAM), lcd_fsensor_state_set);\
+		}\
+		else\
+		{\
+			MENU_ITEM_TOGGLE_P(_T(MSG_FSENSOR), _T(MSG_ON), lcd_fsensor_state_set);\
```
Comment (3d-gussner) [inline Firmware/ultralcd.cpp]: See comment above

File: Firmware/ultralcd.cpp [hunk 5]
```diff
@@ -7118,7 +7141,12 @@ static void lcd_tune_menu()
                MENU_ITEM_TOGGLE_P(_T(MSG_FSENSOR), _T(MSG_OFF), lcd_fsensor_state_set);
           }
 	}
-	else {
+	else if (oFSensorMode==ClFSensorMode::_On_And_Jam && mmu_enabled)
+	{	// Jam detect mode
+       	MENU_ITEM_TOGGLE_P(_T(MSG_FSENSOR), _T(MSG_ON_JAM), lcd_fsensor_state_set);
```
Comment (3d-gussner) [inline Firmware/ultralcd.cpp]: See comment above

**Review by 3d-gussner — COMMENTED** [review]
It would also very helpful if you try to update the `lang_en.txt` and `lang_en_xx.txt` (maybe even the `po`files) files.
Please read [translations.md](https://github.com/prusa3d/Prusa-Firmware/blob/MK3/lang/translations.md)

If you don't feel comfortable with that please let me know and I will create a PR to your PR.

**Comment by vintagepc** [comment]
I will have a look - I intend to break out the jam detection option due to length constraints.

**Comment by vintagepc** [comment]
@3d-gussner Can you clarify the nature of the "c" and "r" values? Is it to document the maximum length used by the biggest language?
For example:
ISTR("Sheet"); ////c=10
means that one of the translations has a 10-character word for "Sheet"?

**Comment by vintagepc** [comment]
I've reworked the menu to be two separate items, though they still share the same EEPROM byte, and pushed up a partial.

image

image

image

I did try to do the language stuff but I think I'm doing something wrong. After doing the lang_add.sh lang_add.txt, it adds them to the text files, but it does not appear to tag them with the correct identifier despite the comments in the code.  AFAICT they are structured similarly to other nearby items so it's not obvious to me what is wrong:
I looked at MSG_REMOVE_OLD_FILAMENT and MSG_FSENSOR_NA as examples. 

example .txt content:
```
#
"Possible extruder jam detected (IR=0). Fix problem and press knob."

```

File: Firmware/messages.c (hunk 1 above)
Comment (vintagepc) [reply Firmware/messages.c]: This entry was removed.

**Comment by DRracer** [comment]
@vintagepc we'd like to postpone this PR after FW 3.10 as there will be some greater care devoted to the MMU itself. Therefore I'm removing the 3.10 tag from this PR.

**Comment by vintagepc** [comment]
I'll hold you to that :stuck_out_tongue:

**Comment by 3d-gussner** [comment]
@vintagepc Please check if this is still a valid PR. If so please rebase update and add a miles stone to it.

**Comment by vintagepc** [comment]
Closing because the code it was based on is obsolete

**Pull request closed without merging**